onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc

```diff
@@ -0,0 +1,129 @@
+        const uint8_t* Input = x + first * kMiniChannelGroup;
+        uint8_t* Output = y + first * kMiniChannelGroup;
+        int64_t channel_count = (last == channel_groups) ? (C - first * kMiniChannelGroup) : ((last - first) * kMiniChannelGroup);
+        MlasNhwcQLinearGlobalAveragePool(
+            Input, x_scale, x_zero_point, Output, y_scale, y_zero_point,
+            N, kernel_size, C, channel_count, acc_buffer.data(), zero_buffer.data());
+      };
+      concurrency::ThreadPool::TryParallelFor(
+          tp, static_cast<std::ptrdiff_t>(channel_groups),
+          {1.0 * N * kernel_size * kMiniChannelGroup, 1.0 * N * kMiniChannelGroup, 8.0 * N * kernel_size * kMiniChannelGroup},
+          worker);
+    } else {
+      auto worker = [=](std::ptrdiff_t first, std::ptrdiff_t last) {
+        const uint8_t* Input = x + first * C * kernel_size;
+        uint8_t* Output = y + first * C;
+        std::vector<int32_t> acc_buffer(MlasQLinearSafePaddingElementCount(sizeof(int32_t), C));
+        std::vector<uint8_t> zero_buffer(MlasQLinearSafePaddingElementCount(sizeof(uint8_t), C), 0);
+        MlasNhwcQLinearGlobalAveragePool(
+            Input, x_scale, x_zero_point, Output, y_scale, y_zero_point,
+            last - first, kernel_size, C, C, acc_buffer.data(), zero_buffer.data());
+      };
+      concurrency::ThreadPool::TryParallelFor(
+          tp, static_cast<std::ptrdiff_t>(N),
+          {1.0 * kernel_size * C, 1.0 * C, 8.0 *kernel_size * C},
+          worker);
+    }
+  }
+  return Status::OK();
+}
+
+Status QLinearGlobalAveragePool::Compute(OpKernelContext* context) const {
+  auto tensor_x_scale = context->Input<Tensor>(1);
+  auto tensor_x_zero_point = context->Input<Tensor>(2);
+  auto tensor_y_scale = context->Input<Tensor>(3);
+  auto tensor_y_zero_point = context->Input<Tensor>(4);
+
+  ORT_ENFORCE(IsScalarOr1ElementVector(tensor_x_scale),
+              "Input x_scale must be a scalar or 1D tensor of size 1");
+  ORT_ENFORCE(IsScalarOr1ElementVector(tensor_x_zero_point),
+              "input x_zero_point must be a scalar or 1D tensor of size 1 if given");
+  ORT_ENFORCE(IsScalarOr1ElementVector(tensor_y_scale),
+              "input y_scale must be a scalar or 1D tensor of size 1");
+  ORT_ENFORCE(IsScalarOr1ElementVector(tensor_y_zero_point),
+              "input y_zero_point must be a scalar or 1D tensor of size 1 if given");
+
+  concurrency::ThreadPool* tp = context->GetOperatorThreadPool();
+  const auto& X = *context->Input<Tensor>(0);
+  const auto& x_shape = X.Shape().GetDims();
+
+  ORT_RETURN_IF_NOT(x_shape.size() >= 3, "Input dimension cannot be less than 3.");
+  auto first_dim = x_shape.begin() + (storage_order_ == StorageOrder::NCHW ? 2 : 1);
+  std::vector<int64_t> kernel_dims{first_dim, first_dim + (x_shape.size() - 2)};
+  int64_t N = x_shape[0];
+  int64_t C = (storage_order_ == StorageOrder::NCHW ? x_shape[1] : x_shape.back());
+
+  std::vector<int64_t> output_dims{N};
```

Comment:
Would the code be smaller to copy x_shape to output_dims and then just set first_dim to first_dim + kernel_rank to 1? One less std::vector to construct plus not all the push_back calls.

Reply:
Fixed.